cloudpathlib/cloudpath.py

```diff
@@ -357,44 +357,66 @@ def _glob_checks(self, pattern: str) -> None:
         if pattern.startswith(self.cloud_prefix) or pattern.startswith("/"):
             raise CloudPathNotImplementedError("Non-relative patterns are unsupported")
 
-    def _glob(self: DerivedCloudPath, selector) -> Generator[DerivedCloudPath, None, None]:
+    def _glob(
+        self: DerivedCloudPath, selector, recursive: bool
+    ) -> Generator[DerivedCloudPath, None, None]:
+        # build a tree structure for all files out of default dicts
+        Tree: Callable = lambda: defaultdict(Tree)
+
+        def _build_tree(trunk, branch, nodes, is_dir):
+            """Utility to build a tree from nested defaultdicts with a generator
+            of nodes (parts) of a path."""
+            next_branch = next(nodes, None)
+
+            if next_branch is None:
+                trunk[branch] = Tree() if is_dir else None  # leaf node
+
+            else:
+                _build_tree(trunk[branch], next_branch, nodes, is_dir)
+
+        file_tree = Tree()
+
+        for f, is_dir in self.client._list_dir(self, recursive=recursive):
+            parts = str(f.relative_to(self)).split("/")
+
+            # skip self
+            if len(parts) == 1 and parts[0] == ".":
+                continue
+
+            nodes = (p for p in parts)
+            _build_tree(file_tree, next(nodes, None), nodes, is_dir)
+
+        file_tree = dict(file_tree)  # freeze as normal dict before passing in
+
         root = _CloudPathSelectable(
-            PurePosixPath(self._no_prefix_no_drive),
-            {
-                PurePosixPath(c._no_prefix_no_drive): is_dir
-                for c, is_dir in self.client._list_dir(self, recursive=True)
-            },
-            is_dir=True,
-            exists=True,
+            self.name,
+            [p.name for p in self.parents[:-1]],  # all parents except bucket/container
+            file_tree,
         )
 
         for p in selector.select_from(root):
-            yield self.client.CloudPath(f"{self.cloud_prefix}{self.drive}{p}")
+            yield self.client.CloudPath(f"{self.cloud_prefix}{self.drive}/{p}")
 
     def glob(self: DerivedCloudPath, pattern: str) -> Generator[DerivedCloudPath, None, None]:
-        if pattern == "*":
-            yield from (path for path, is_file in self.client._list_dir(self, recursive=False))
-            return
-        if pattern == "**/*":
-            yield from (path for path, is_file in self.client._list_dir(self, recursive=True))
-            return
         self._glob_checks(pattern)
 
         pattern_parts = PurePosixPath(pattern).parts
         selector = _make_selector(tuple(pattern_parts), _posix_flavour)
 
-        yield from self._glob(selector)
+        yield from self._glob(
+            selector,
+            "/" in pattern
+            or "**"
+            in pattern,  # recursive listing needed if explicit ** or any sub folder in pattern
+        )
```

Comment:
Some food for thought: Wondering if this is still searching a larger part of the file tree than needed.

If you have `Cloudpath("s3://bucket/depth1").glob("depth2/depth3/*.txt")`, it seems like this will recursively search `s3://bucket/depth1` (i.e., `self`). But what we really care about is a non-recursive listing of `s3://bucket/depth1/depth2/depth3`. If any of the parent directories have sibling directories with a lot of files, then I think this ends up being more expensive than necessary? Maybe a further optimization to, rather than doing `self._glob`, construct a `CloudPath` from all pattern parts up until one with a `*` in it? 

Maybe this is an unnecessary optimization. Or maybe there's something about how pathlib's glob works that makes this not make sense—I don't really remember the details about what was wrong with our old glob.

Reply:
This is a good point, and I started to think about optimizations, but got worried that I'd miss some idiosyncrasy about glob logic or re-implement too much of it naïvely.

To me, I think the current overhead seems pretty acceptable, but I'll file an issue from this to see if anyone has glob perf complaints after this and we want to optimize more.